File: frontend/js/views/dividend_discount_model.js
// Dividend discount model — Gordon growth or two-stage intrinsic value, via
// /calc/dividend-discount-model.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderDividendDiscountModel(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.ddm.h1.title">// DIVIDEND DISCOUNT MODEL</span></h1>
        <p class="muted small" data-i18n="view.ddm.hint.intro">
            Intrinsic share value as the present value of future dividends. The Gordon model assumes constant
            growth in perpetuity; the two-stage model uses a high-growth phase for N years then a perpetual
            terminal growth. Requires the discount rate to exceed the perpetual growth rate.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.ddm.h2.inputs">Model inputs</h2>
            <form id="ddm-form" class="inline-form">
                <label><span data-i18n="view.ddm.label.ticker">Ticker</span>
                    <input type="text" name="ticker" value="ACME"></label>
                <label><span data-i18n="view.ddm.label.model">Model</span>
                    <select name="model">
                        <option value="gordon" data-i18n="view.ddm.opt.gordon">Gordon (constant)</option>
                        <option value="two_stage" data-i18n="view.ddm.opt.two">Two-stage</option>
                    </select></label>
                <label><span data-i18n="view.ddm.label.d0">Current dividend D0 ($)</span>
                    <input type="number" step="0.01" min="0" name="current_dividend_usd" value="2" required></label>
                <label><span data-i18n="view.ddm.label.r">Required return (%)</span>
                    <input type="number" step="0.1" name="required_return_pct" value="9" required></label>
                <label><span data-i18n="view.ddm.label.g">Growth (%)</span>
                    <input type="number" step="0.1" name="growth_pct" value="4" required></label>
                <label><span data-i18n="view.ddm.label.years">High-growth years (two-stage)</span>
                    <input type="number" step="1" min="0" name="high_growth_years" value="5"></label>
                <label><span data-i18n="view.ddm.label.gt">Terminal growth (%, two-stage)</span>
                    <input type="number" step="0.1" name="terminal_growth_pct" value="4"></label>
            </form>
        </div>
        <div id="ddm-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#ddm-form');
    const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const generate = async () => {
        const body = {
            ticker: (form.querySelector('[name="ticker"]').value || '').trim(),
            current_dividend_usd: n('current_dividend_usd'),
            required_return_pct: n('required_return_pct'),
            model: form.querySelector('[name="model"]').value,
            growth_pct: n('growth_pct'),
            high_growth_years: n('high_growth_years'),
            terminal_growth_pct: n('terminal_growth_pct'),
        };
        try {
            const doc = await api.calcDividendDiscount(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.ddm.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#ddm-result');
    if (!doc.valid) {
        el.innerHTML = `<p class="muted" data-i18n="view.ddm.invalid">Invalid — the discount rate must exceed the perpetual growth rate.</p>`;
        applyUiI18n(el);
        return;
    }
    const twoStage = doc.pv_dividends_usd > 0;
    const split = twoStage
        ? `<div class="card"><div class="label" data-i18n="view.ddm.card.pvdiv">PV of dividends</div>
               <div class="value">${money(doc.pv_dividends_usd)}</div></div>
           <div class="card"><div class="label" data-i18n="view.ddm.card.pvterm">PV of terminal</div>
               <div class="value">${money(doc.pv_terminal_usd)}</div></div>`
        : '';
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.ddm.card.fair">Fair value</div>
                    <div class="value">${money(doc.fair_value_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.ddm.card.d1">Next dividend D1</div>
                    <div class="value">${money(doc.next_dividend_usd)}</div></div>
                ${split}
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="ddm-copy" type="button" data-i18n="view.ddm.btn.copy">Copy</button>
            </div>
        </div>
        <p class="muted small">${esc(doc.model_label)}</p>
    `;
    applyUiI18n(el);
    el.querySelector('#ddm-copy').addEventListener('click', async () => {
        const txt = `${doc.model_label} fair value: ${money(doc.fair_value_usd)} (D1 ${money(doc.next_dividend_usd)})`;
        try { await navigator.clipboard.writeText(txt); showToast(t('view.ddm.toast.copied'), { level: 'success' }); }
        catch (e) { showToast(t('view.ddm.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' }); }
    });
}

File: crates/traderview-core/src/dividend_discount_model.rs
//! Dividend discount model (DDM) — intrinsic share value as the present value of
//! future dividends. Supports the single-stage Gordon growth model (constant
//! growth in perpetuity) and a two-stage model (a high-growth phase for N years,
//! then a perpetual terminal growth). Reports the fair value and, for two-stage,
//! the split between the explicit dividends and the terminal value. Distinct from
//! the implied-dividend module (which backs dividend yield out of option prices).
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct DdmInput {
    pub ticker: String,
    /// Most recent annual dividend per share (D0).
    pub current_dividend_usd: f64,
    /// Required rate of return (discount rate), percent.
    pub required_return_pct: f64,
    /// "gordon" (constant growth) or "two_stage".
    #[serde(default)]
    pub model: String,
    /// Constant growth (Gordon) or high-growth-phase rate (two-stage), percent.
    pub growth_pct: f64,
    /// High-growth phase length in years (two-stage only).
    #[serde(default)]
    pub high_growth_years: u32,
    /// Terminal perpetual growth rate, percent (two-stage only).
    #[serde(default)]
    pub terminal_growth_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct DdmReport {
    pub model_label: String,
    /// Intrinsic value per share (0 if the model is invalid, e.g. g ≥ r).
    pub fair_value_usd: f64,
    /// Present value of the explicit-phase dividends (two-stage; equals 0 for Gordon).
    pub pv_dividends_usd: f64,
    /// Present value of the terminal value (two-stage; equals the full value for Gordon).
    pub pv_terminal_usd: f64,
    /// Next-year dividend D1 = D0 × (1 + growth).
    pub next_dividend_usd: f64,
    pub valid: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

pub fn generate(i: &DdmInput) -> DdmReport {
    let r = i.required_return_pct / 100.0;
    let g = i.growth_pct / 100.0;
    let d0 = i.current_dividend_usd;
    let two_stage = i.model.trim().eq_ignore_ascii_case("two_stage");

    if two_stage {
        let g2 = i.terminal_growth_pct / 100.0;
        let n = i.high_growth_years;
        // Terminal perpetuity requires r > terminal growth.
        if r <= g2 || n == 0 {
            return DdmReport { model_label: "Two-stage".into(), ..Default::default() };
        }
        let mut pv_div = 0.0;
        let mut d = d0;
        for t in 1..=n {
            d *= 1.0 + g;
            pv_div += d / (1.0 + r).powi(t as i32);
        }
        // d now holds D_n; terminal value at year n uses the terminal growth.
        let terminal = d * (1.0 + g2) / (r - g2);
        let pv_terminal = terminal / (1.0 + r).powi(n as i32);
        let fair = pv_div + pv_terminal;
        DdmReport {
            model_label: "Two-stage".into(),
            fair_value_usd: round2(fair),
            pv_dividends_usd: round2(pv_div),
            pv_terminal_usd: round2(pv_terminal),
            next_dividend_usd: round2(d0 * (1.0 + g)),
            valid: true,
        }
    } else {
        // Gordon growth requires r > g.
        if r <= g {
            return DdmReport { model_label: "Gordon growth".into(), ..Default::default() };
        }
        let d1 = d0 * (1.0 + g);
        let fair = d1 / (r - g);
        DdmReport {
            model_label: "Gordon growth".into(),
            fair_value_usd: round2(fair),
            pv_dividends_usd: 0.0,
            pv_terminal_usd: round2(fair),
            next_dividend_usd: round2(d1),
            valid: true,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.02
    }

    #[test]
    fn gordon_growth() {
        let d = generate(&DdmInput {
            ticker: "ACME".into(),
            current_dividend_usd: 2.0,
            required_return_pct: 9.0,
            model: "gordon".into(),
            growth_pct: 4.0,
            high_growth_years: 0,
            terminal_growth_pct: 0.0,
        });
        assert!(d.valid);
        assert!(close(d.fair_value_usd, 41.60));
        assert!(close(d.next_dividend_usd, 2.08));
    }

    #[test]
    fn two_stage() {
        let d = generate(&DdmInput {
            ticker: "ACME".into(),
            current_dividend_usd: 2.0,
            required_return_pct: 9.0,
            model: "two_stage".into(),
            growth_pct: 15.0,
            high_growth_years: 5,
            terminal_growth_pct: 4.0,
        });
        assert!(d.valid);
        assert!(close(d.pv_dividends_usd, 11.78));
        assert!(close(d.pv_terminal_usd, 54.38));
        assert!(close(d.fair_value_usd, 66.16));
    }

    #[test]
    fn gordon_invalid_when_growth_exceeds_return() {
        let d = generate(&DdmInput {
            ticker: "X".into(),
            current_dividend_usd: 2.0,
            required_return_pct: 5.0,
            model: "gordon".into(),
            growth_pct: 6.0,
            high_growth_years: 0,
            terminal_growth_pct: 0.0,
        });
        assert!(!d.valid);
        assert!(close(d.fair_value_usd, 0.0));
    }

    #[test]
    fn two_stage_invalid_terminal_growth() {
        let d = generate(&DdmInput {
            ticker: "X".into(),
            current_dividend_usd: 2.0,
            required_return_pct: 5.0,
            model: "two_stage".into(),
            growth_pct: 10.0,
            high_growth_years: 5,
            terminal_growth_pct: 6.0,
        });
        assert!(!d.valid);
    }

    #[test]
    fn two_stage_splits_sum_to_fair_value() {
        let d = generate(&DdmInput {
            ticker: "ACME".into(),
            current_dividend_usd: 3.0,
            required_return_pct: 10.0,
            model: "two_stage".into(),
            growth_pct: 12.0,
            high_growth_years: 7,
            terminal_growth_pct: 3.0,
        });
        assert!(close(d.fair_value_usd, d.pv_dividends_usd + d.pv_terminal_usd));
    }
}
